app/src/org/commcare/android/util/MediaUtil.java

```diff
@@ -1,82 +1,297 @@
+                                                     int containerWidth) {
+        // Determine dimensions of original image
+        BitmapFactory.Options o = new BitmapFactory.Options();
+        o.inJustDecodeBounds = true;
+        BitmapFactory.decodeFile(imageFilepath, o);
+        int imageHeight = o.outHeight;
+        int imageWidth = o.outWidth;
+
+        return getBitmapScaledByTargetAndContainer(imageFilepath, imageHeight, imageWidth, -1, -1,
+                containerHeight, containerWidth);
+    }
+
+    public static Bitmap getBitmapScaledToContainer(File imageFile, int containerHeight,
+                                                    int containerWidth) {
+        return getBitmapScaledToContainer(imageFile.getAbsolutePath(), containerHeight,
+                containerWidth);
+    }
+
+    /**
+     * @return A bitmap representation of the given image file, scaled down such that the new
+     * dimensions of the image are the SMALLER of the following 2 options:
+     * 1) targetHeight and targetWidth
+     * 2) the largest dimensions for which the original aspect ratio is maintained, without
+     * exceeding either boundingWidth or boundingHeight
+     *
+     * Provides for the possibility that there is no target height or target width (indicated by
+     * setting them to -1), in which case the 2nd option above is used.
+     */
+    private static Bitmap getBitmapScaledByTargetAndContainer(String imageFilepath,
+                                                   int originalHeight, int originalWidth,
+                                                   int targetHeight, int targetWidth,
+                                                   int boundingHeight, int boundingWidth) {
+
+        Pair<Integer, Integer> dimensImposedByContainer = getProportionalDimensForContainer(
+                originalHeight, originalWidth, boundingHeight, boundingWidth);
+
+        int newWidth, newHeight;
+        if (targetHeight == -1 || targetWidth == -1) {
+            newWidth = dimensImposedByContainer.first;
+            newHeight = dimensImposedByContainer.second;
+        } else {
+            newWidth = Math.min(dimensImposedByContainer.first, targetWidth);
+            newHeight = Math.min(dimensImposedByContainer.second, targetHeight);
+        }
+
+        int approximateScaleFactor = originalWidth / newWidth;
+        try {
+            BitmapFactory.Options o = new BitmapFactory.Options();
+            o.inSampleSize = approximateScaleFactor;
+            // Decode the bitmap with the largest integer scale down factor that will not make it
+            // smaller than the final desired size
+            Bitmap originalBitmap = BitmapFactory.decodeFile(imageFilepath, o);
+            // From that, generate a bitmap scaled to the exact right dimensions
+            return Bitmap.createScaledBitmap(originalBitmap, newWidth, newHeight, false);

```

Comment:
So when we aren't scaling for density reasons (where the exact final dimension isn't required to be specific), I'm worried about the runtime implications of this scaling. If we load an image and don't density-adjust it and it's 20% bigger than it needs to be it seems like this is gonna really slow things down, right? I can't imagine it's a fast operation. 

(Just trying to consider implications for people who don't have the new feature enabled)